File: android-app/app/src/main/java/com/plantgotchi/app/ble/SensorParser.kt
package com.plantgotchi.app.ble

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Parses raw BLE characteristic byte arrays into sensor values.
 *
 * All multi-byte values are little-endian per BLE GATT convention.
 */
object SensorParser {

    /**
     * Parse moisture from a 2-byte uint16 value.
     * @return moisture percentage (0-100), or null if data is invalid.
     */
    fun parseMoisture(data: ByteArray): Int? {
        if (data.size < 2) return null
        val value = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN).short.toInt() and 0xFFFF
        return value.coerceIn(0, 100)
    }

    /**
     * Parse temperature from a 2-byte int16 value (degrees C * 100).
     * @return temperature in degrees Celsius, or null if data is invalid.
     */
    fun parseTemperature(data: ByteArray): Double? {
        if (data.size < 2) return null
        val raw = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN).short.toInt()
        return raw / 100.0
    }

    /**
     * Parse light from a 4-byte uint32 value.
     * @return light level in lux, or null if data is invalid.
     */
    fun parseLight(data: ByteArray): Int? {
        if (data.size < 4) return null
        val value = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN).int.toLong() and 0xFFFFFFFFL
        return value.toInt()
    }

    /**
     * Parse battery from a 1-byte uint8 value.
     * @return battery percentage (0-100), or null if data is invalid.
     */
    fun parseBattery(data: ByteArray): Int? {
        if (data.isEmpty()) return null
        val value = data[0].toInt() and 0xFF
        return value.coerceIn(0, 100)
    }
}

File: ios-app/Plantgotchi/BLE/SensorParser.swift
import Foundation

/// Parses raw BLE characteristic `Data` into typed sensor values.
/// The ESP32 sensor sends integer-encoded values (no floats over BLE).
enum SensorParser {

    /// Parse moisture from uint16 little-endian data. Returns percentage 0-100.
    static func parseMoisture(_ data: Data) -> Int? {
        guard data.count >= 2 else { return nil }
        let value = data.withUnsafeBytes { $0.load(as: UInt16.self) }
        let moisture = Int(UInt16(littleEndian: value))
        guard moisture >= 0, moisture <= 100 else { return nil }
        return moisture
    }

    /// Parse temperature from int16 little-endian data. Value is degrees C x 100.
    /// Returns temperature in degrees Celsius as a Double.
    static func parseTemperature(_ data: Data) -> Double? {
        guard data.count >= 2 else { return nil }
        let value = data.withUnsafeBytes { $0.load(as: Int16.self) }
        let raw = Int16(littleEndian: value)
        return Double(raw) / 100.0
    }

    /// Parse light from uint32 little-endian data. Returns lux value.
    static func parseLight(_ data: Data) -> Int? {
        guard data.count >= 4 else { return nil }
        let value = data.withUnsafeBytes { $0.load(as: UInt32.self) }
        return Int(UInt32(littleEndian: value))
    }

    /// Parse battery from uint8 data. Returns percentage 0-100.
    static func parseBattery(_ data: Data) -> Int? {
        guard data.count >= 1 else { return nil }
        let value = Int(data[0])
        guard value >= 0, value <= 100 else { return nil }
        return value
    }
}
